File: counting-sort.js
/*
 * Couring sort..
 * O(n + k)
 * k is the range
 */
let range = 255;

let coutingSort = function(arr) {
  let count = new Array(range);
  count.fill(0);
  let output = new Array(arr.length);
  output.fill(0);

  arr.forEach(function(val) {
    ++count[val];
  });

  for (let j = 1; j < range; ++j) {
    count[j] += count[j - 1];
  }

  arr.forEach(function(val) {
    output[count[val] - 1] = val;
    --count[val];
  })

  return output;
};

console.log(coutingSort([12, 1, 5, 4, 4, 0, 3, 9, 3, 4]));

File: counting_sort.cc
/* Nov 23, 2018
 * Counting sort..
 */

#include <iostream>
#include <vector>

#define RANGE 255

using namespace std;

vector<int> couting_sort(vector<int>& arr) {
  vector<int> result(arr.size(), 0);
  vector<int> count(RANGE + 1, 0);

  for (auto a : arr) {
    ++count[a];
  }

  for (int i = 1; i <= RANGE; ++i) {
    count[i] += count[i - 1];
    // To get cummulative position
  }

  for (auto a : arr) {
    result[count[a] - 1] = a;
    --count[a];
  }

  return result;
}

int main() {
  vector<int> arr{12, 1, 5, 4, 4, 0, 3, 9, 3, 4};
  auto re = couting_sort(arr);
  for (auto r : re) {
    cout << r << " ";
  }
  cout << endl;

  return 0;
}
